File: crates/harness-tools/src/ask.rs
//! Text-based human-in-the-loop request tool.
//!
//! `ask.text` is intentionally the single text transport entry point:
//! it can render as confirmation, free-form input, or a short choice
//! list through its `kind` argument. Future modalities can live beside
//! it as `ask.voice`, `ask.video`, etc. while reusing the same native
//! HITL request/response protocol.

use async_trait::async_trait;
use harness_core::{
    request_human, BoxError, HitlKind, HitlOption, HitlRequest, HitlResponse, HitlTransport, Tool,
    ToolCategory,
};
use serde_json::{json, Value};

pub struct AskTextTool;

#[async_trait]
impl Tool for AskTextTool {
    fn name(&self) -> &str {
        "ask.text"
    }

    fn category(&self) -> ToolCategory {
        ToolCategory::Read
    }

    fn description(&self) -> &str {
        "Ask the human operator a text-based question. Use when you need missing information, a confirmation, or a choice instead of guessing."
    }

    fn parameters(&self) -> Value {
        json!({
            "type": "object",
            "properties": {
                "kind": {
                    "type": "string",
                    "enum": ["input", "confirm", "choice", "review"],
                    "description": "How the text prompt should be rendered. Defaults to input."
                },
                "title": {
                    "type": "string",
                    "description": "Short prompt title."
                },
                "body": {
                    "type": "string",
                    "description": "Optional context, proposed action, or question details."
                },
                "options": {
                    "type": "array",
                    "items": { "type": "string" },
                    "description": "Required for kind=choice. Each string is used as both label and value."
                },
                "default_value": {
                    "type": "string",
                    "description": "Optional prefilled text or default choice value."
                },
                "multiline": {
                    "type": "boolean",
                    "description": "For kind=input, whether a multiline editor is preferred. Defaults to true."
                }
            },
            "required": ["title"],
            "additionalProperties": false
        })
    }

    async fn invoke(&self, args: Value) -> Result<String, BoxError> {
        let title = required_string(&args, "title")?;
        let kind = parse_kind(optional_string(&args, "kind").as_deref())?;
        let mut req = HitlRequest::new(kind, title);
        req.transport = HitlTransport::Text;
        req.body = optional_string(&args, "body");
        req.default_value = optional_string(&args, "default_value").map(Value::String);
        req.metadata = Some(json!({
            "tool": "ask.text",
            "multiline": args.get("multiline").and_then(Value::as_bool).unwrap_or(true)
        }));

        if matches!(kind, HitlKind::Choice) {
            req.options = parse_options(&args)?;
        }

        response_json(request_human(req).await?)
    }
}

fn parse_kind(raw: Option<&str>) -> Result<HitlKind, BoxError> {
    match raw.unwrap_or("input") {
        "input" => Ok(HitlKind::Input),
        "confirm" => Ok(HitlKind::Confirm),
        "choice" => Ok(HitlKind::Choice),
        "review" => Ok(HitlKind::Review),
        other => Err(format!("unsupported ask.text kind `{other}`").into()),
    }
}

fn parse_options(args: &Value) -> Result<Vec<HitlOption>, BoxError> {
    let options = args
        .get("options")
        .and_then(Value::as_array)
        .ok_or_else(|| -> BoxError { "`options` must be an array for kind=choice".into() })?;
    let options = options
        .iter()
        .map(|v| {
            let s = v
                .as_str()
                .ok_or_else(|| -> BoxError { "every option must be a string".into() })?;
            Ok(HitlOption {
                value: s.to_string(),
                label: s.to_string(),
            })
        })
        .collect::<Result<Vec<_>, BoxError>>()?;
    if options.is_empty() {
        return Err("`options` must not be empty for kind=choice".into());
    }
    Ok(options)
}

fn response_json(response: HitlResponse) -> Result<String, BoxError> {
    serde_json::to_string(&response).map_err(Into::into)
}

fn required_string(args: &Value, key: &str) -> Result<String, BoxError> {
    args.get(key)
        .and_then(Value::as_str)
        .filter(|s| !s.trim().is_empty())
        .map(ToString::to_string)
        .ok_or_else(|| format!("`{key}` must be a non-empty string").into())
}

fn optional_string(args: &Value, key: &str) -> Option<String> {
    args.get(key)
        .and_then(Value::as_str)
        .map(str::trim)
        .filter(|s| !s.is_empty())
        .map(ToString::to_string)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ask_text_defaults_to_input() {
        assert_eq!(parse_kind(None).unwrap(), HitlKind::Input);
    }

    #[test]
    fn choice_requires_options() {
        let err = parse_options(&json!({ "options": [] })).unwrap_err();
        assert!(err.to_string().contains("must not be empty"));
    }

    #[test]
    fn schema_requires_only_title() {
        let tool = AskTextTool;
        let schema = tool.parameters();
        assert_eq!(schema["required"], json!(["title"]));
    }
}

File: packages/tools/src/ask.ts
// `ask.text` — text-based human-in-the-loop request tool.
// Ported from harness-tools/src/ask.rs.
//
// `ask.text` is the single text transport entry point: it can render as
// confirmation, free-form input, or a short choice list through its `kind`
// argument. Future modalities can live beside it as `ask.voice`, `ask.video`,
// etc. while reusing the same native HITL request/response protocol.
//
// DEFERRAL: the Rust version calls `harness_core::request_human(req)` which
// publishes a PendingHitl over a task-local mpsc channel installed by the
// agent loop / transport, then awaits the human's HitlResponse. @jarvis/core
// does not yet expose a HITL channel (no `request_human` / `with_hitl`), so
// this port is best-effort: it does the full arg validation and builds the
// HitlRequest, but — having no transport to await — returns the request as
// JSON tagged `"deferred": "no HITL channel"` so the agent loop has a textual
// result and the clarification text is surfaced. Swap the body of #dispatch
// for a real `requestHuman(req)` await once core lands the channel.
import { randomUUID } from "node:crypto";
import type { JsonValue } from "@jarvis/core";
import type { Tool, ToolCategory } from "@jarvis/core";

export type HitlKind = "confirm" | "input" | "choice" | "review";

export interface HitlOption {
  value: string;
  label: string;
}

export interface HitlRequest {
  id: string;
  transport: "text" | "voice" | "video";
  kind: HitlKind;
  title: string;
  body?: string;
  options?: HitlOption[];
  default_value?: JsonValue;
  metadata?: JsonValue;
}

// Ids must be collision-proof across process restarts: the web store dedups
// incoming HITL cards on `request.id` and retains cancelled cards, so a
// process-global counter that resets to 1 on restart could re-mint an id that
// collides with a stale cancelled card, silently dropping the new card and
// hanging the turn on `human.request()`. A random UUID never collides.
function mintId(): string {
  return `hitl_${randomUUID()}`;
}

function asObject(args: JsonValue): { [k: string]: JsonValue } {
  if (args === null || typeof args !== "object" || Array.isArray(args)) {
    return {};
  }
  return args;
}

function parseKind(raw: string | undefined): HitlKind {
  switch (raw ?? "input") {
    case "input":
      return "input";
    case "confirm":
      return "confirm";
    case "choice":
      return "choice";
    case "review":
      return "review";
    default:
      throw new Error(`unsupported ask.text kind \`${raw}\``);
  }
}

function parseOptions(obj: { [k: string]: JsonValue }): HitlOption[] {
  const raw = obj["options"];
  if (!Array.isArray(raw)) {
    throw new Error("`options` must be an array for kind=choice");
  }
  const options: HitlOption[] = raw.map((v) => {
    if (typeof v !== "string") {
      throw new Error("every option must be a string");
    }
    return { value: v, label: v };
  });
  if (options.length === 0) {
    throw new Error("`options` must not be empty for kind=choice");
  }
  return options;
}

// `title` (and other required strings): present, a string, and non-empty after
// trimming.
function requiredString(obj: { [k: string]: JsonValue }, key: string): string {
  const v = obj[key];
  if (typeof v === "string" && v.trim().length > 0) {
    return v;
  }
  throw new Error(`\`${key}\` must be a non-empty string`);
}

// Optional string: trimmed; empty / missing / non-string → undefined.
function optionalString(obj: { [k: string]: JsonValue }, key: string): string | undefined {
  const v = obj[key];
  if (typeof v !== "string") return undefined;
  const trimmed = v.trim();
  return trimmed.length > 0 ? trimmed : undefined;
}

export class AskTextTool implements Tool {
  readonly name = "ask.text";
  readonly category: ToolCategory = "read";
  readonly description =
    "Ask the human operator a text-based question. Use when you need missing " +
    "information, a confirmation, or a choice instead of guessing.";

  readonly parameters: JsonValue = {
    type: "object",
    properties: {
      kind: {
        type: "string",
        enum: ["input", "confirm", "choice", "review"],
        description: "How the text prompt should be rendered. Defaults to input.",
      },
      title: {
        type: "string",
        description: "Short prompt title.",
      },
      body: {
        type: "string",
        description: "Optional context, proposed action, or question details.",
      },
      options: {
        type: "array",
        items: { type: "string" },
        description:
          "Required for kind=choice. Each string is used as both label and value.",
      },
      default_value: {
        type: "string",
        description: "Optional prefilled text or default choice value.",
      },
      multiline: {
        type: "boolean",
        description:
          "For kind=input, whether a multiline editor is preferred. Defaults to true.",
      },
    },
    required: ["title"],
    additionalProperties: false,
  };

  async invoke(args: JsonValue): Promise<string> {
    const obj = asObject(args);
    const title = requiredString(obj, "title");
    const kind = parseKind(optionalString(obj, "kind"));

    const multiline =
      typeof obj["multiline"] === "boolean" ? (obj["multiline"] as boolean) : true;

    const req: HitlRequest = {
      id: mintId(),
      transport: "text",
      kind,
      title,
      metadata: { tool: "ask.text", multiline },
    };

    const body = optionalString(obj, "body");
    if (body !== undefined) req.body = body;

    const defaultValue = optionalString(obj, "default_value");
    if (defaultValue !== undefined) req.default_value = defaultValue;

    if (kind === "choice") {
      req.options = parseOptions(obj);
    }

    return this.#dispatch(req);
  }

  // DEFERRED dispatch: no HITL transport in @jarvis/core yet. Return the built
  // request as JSON so the clarification text reaches the model. Replace with
  // `JSON.stringify(await requestHuman(req))` once core exposes the channel.
  #dispatch(req: HitlRequest): Promise<string> {
    return Promise.resolve(JSON.stringify({ deferred: "no HITL channel", request: req }));
  }
}
